File: src/main/java/skytils/skytilsmod/features/impl/misc/ArmorColor.java
package skytils.skytilsmod.features.impl.misc;

import com.google.gson.*;
import net.minecraft.client.Minecraft;
import skytils.skytilsmod.Skytils;

import java.awt.*;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.HashMap;
import java.util.Map;

public class ArmorColor {

    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private static final Minecraft mc = Minecraft.getMinecraft();
    private static File colorFile;

    public static final HashMap<String, Color> armorColors = new HashMap<>();


    public ArmorColor() {
        colorFile = new File(Skytils.modDir, "armorcolors.json");
        reloadColors();
    }

    public static void reloadColors() {
        armorColors.clear();
        JsonObject dataObject;
        try (FileReader in = new FileReader(colorFile)) {
            dataObject = gson.fromJson(in, JsonObject.class);
            for (Map.Entry<String, JsonElement> colors : dataObject.entrySet()) {
                armorColors.put(colors.getKey(), Color.decode(colors.getValue().getAsString()));
            }
        } catch (Exception e) {
            dataObject = new JsonObject();
            try (FileWriter writer = new FileWriter(colorFile)) {
                gson.toJson(dataObject, writer);
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    public static void saveColors() {
        try (FileWriter writer = new FileWriter(colorFile)) {
            JsonObject obj = new JsonObject();
            for (Map.Entry<String, Color> colors : armorColors.entrySet()) {
                obj.addProperty(colors.getKey(), String.format("#%06X", (0xFFFFFF & colors.getValue().getRGB())));
            }
            gson.toJson(obj, writer);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

}

File: src/main/kotlin/sharttils/sharttilsmod/features/impl/handlers/ArmorColor.kt
/*
 * Sharttils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2022 Sharttils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package sharttils.sharttilsmod.features.impl.handlers

import com.google.gson.JsonObject
import sharttils.sharttilsmod.Sharttils
import sharttils.sharttilsmod.core.PersistentSave
import sharttils.sharttilsmod.utils.Utils
import sharttils.sharttilsmod.utils.graphics.colors.CustomColor
import java.io.*

class ArmorColor : PersistentSave(File(Sharttils.modDir, "armorcolors.json")) {

    override fun read(reader: InputStreamReader) {
        armorColors.clear()
        val dataObject = gson.fromJson(reader, JsonObject::class.java)
        for ((key, value) in dataObject.entrySet()) {
            val color = Utils.customColorFromString(value.asString)
            armorColors[key] = color
        }
    }

    override fun write(writer: OutputStreamWriter) {
        val obj = JsonObject()
        for ((key, value) in armorColors) {
            obj.addProperty(key, value.toString())
        }
        gson.toJson(obj, writer)
    }

    override fun setDefault(writer: OutputStreamWriter) {
        gson.toJson(JsonObject(), writer)
    }

    companion object {
        @JvmField
        val armorColors = HashMap<String, CustomColor>()
    }
}